x/wasm/watcher/write.go

```diff
@@ -0,0 +1,80 @@
+package watcher
+
+import (
+	"log"
+
+	sdk "github.com/okex/exchain/libs/cosmos-sdk/types"
+)
+
+func NewHeight() {
+	if Enable() && len(blockStateCache) != 0 {
+		blockStateCache = make(map[string]*WatchMessage)
+	}
+}
+
+func Save(err error) {
+	if !Enable() {
+		return
+	}
+	txCacheMtx.Lock()
+	if err == nil {
+		for _, msg := range txStateCache {
+			blockStateCache[string(msg.Key)] = msg
+		}
+	}
+	txStateCache = txStateCache[:0]
+	txCacheMtx.Unlock()
+}
+
+func Commit() {
+	if !Enable() {
+		return
+	}
+	blockStateCacheCopy := blockStateCache
```

Comment:
It's not thread safe. Commit is async, when save and commit using the same slice,will panic